### PR TITLE
[Distributed Strategy] Support HySparse SWA KV cache for generation

### PR DESCRIPTION
**Pull request by xiaoguoguo626807**
### PR Category

Execute Infrastructure

### PR Types

Improvements

### Description

适配Hysparse swa kv cache 更新

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/multi_latent_attention.py [hunk 1]
```diff
@@ -1950,73 +2030,87 @@ def compute_absorbed_v(core_attn_out):
         # =================
         # Sparse attention computation
         # =================
+        # Skipped while decoding: block scoring needs the full attention logits
+        # of a whole query row, which the full layer cannot produce one token at
+        # a time. The sliding-window branch above already covers the local
+        # dependency, so decode runs window-only.
+        sparse_core_attn_out = None
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_latent_attention.py]: P0 缓存解码会静默丢弃 HySparse 的 block-sparse 输出。

这里在 `is_decode=True` 时把 `sparse_core_attn_out` 保持为 `None`，后面的加和也随之跳过；但 `no_cache` 生成每步按完整序列前向，仍会计算并加入该分支。因此从第二个生成 token 起，SWA 层不再使用窗口外的 top-k block，缓存与无缓存路径会产生不同 hidden states/logits，这也与 PR 描述中的“精度不变”不一致。当前测试只校验有限值及 full 层一致性，没有覆盖这项差异。

请在 decode 时基于当前 query 和 full-layer KV cache 生成本 token 的 block indices，并用累计的 `shared_key` 执行 sparse 分支；若暂不支持，应对 HySparse cache 明确报错或回退到 `no_cache`，避免静默改变模型语义。同时补充序列长度大于 window 的 full+SWA cache/no-cache 输出对齐测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [multi_latent_attention.py:2037](https://github.com/PaddlePaddle/PaddleFleet/pull/1592#discussion_r3682218932) | P0 | ✅ |
| 2 | [PR 标题](#pr-title) | P3 | ✅ |
| 3 | [multi_latent_attention.py:2114](https://github.com/PaddlePaddle/PaddleFleet/pull/1592#discussion_r3688810349) | P1 | ✅ |

<a id="pr-title"></a>
P3 标题已更新为 `[Distributed Strategy] Support HySparse SWA KV cache for generation`，符合仓库约定的 `[Tag]` 格式并准确描述了变更。本项已解决。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/multi_latent_attention.py [hunk 2]
```diff
@@ -1950,8 +2108,25 @@ def compute_absorbed_v(core_attn_out):
...
+                "were provided."
+            )
+        if is_decode:
+            # The shared latent is never window-truncated, so its length is the
+            # whole KV history; the decode query's document range covers all of
+            # it (generation runs one document per batch row).
+            doc_valid_range = paddle.concat(
+                [
+                    paddle.zeros([b, 1, 1], dtype="int32"),
+                    paddle.full([b, 1, 1], shared_key.shape[1], dtype="int32"),
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_latent_attention.py:2317]: P1 CP decode 在 all-gather 前按本 rank 的 `shared_key.shape[1]` 构造 `doc_valid_range`，但下一段会把 `shared_key` 拼成全局序列后交给 block-sparse kernel。这样 `eos` 仍是 local length，全局历史后半段会被 `valid_range` 屏蔽；同时 full 层生成的 block indices 以 cache 的全局坐标为准，二者坐标不一致，CP decode 的输出会错误。请在 all-gather 后按全局长度重建 range，并统一 cache/index 的 CP 布局，或明确禁止 CP KV-cache decode；补充至少一个多卡 CP decode 对齐测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/multi_latent_attention.py (hunk 1 above)
Comment (xiaoguoguo626807) [reply src/paddlefleet/transformer/multi_latent_attention.py]: done

Comment (risemeup1111) [reply src/paddlefleet/transformer/multi_latent_attention.py]: 已核对当前 head：full 层 decode 通过 `_hy_sparse_decode_block_indices` 从 KV cache 重新计算 `shared_block_indices`，SWA/MQA decode 也重新执行 block-sparse 分支，并新增了 cache/no-cache 对齐测试。因此原 P0 已修复，Review Board 已标记为 ✅。

File: src/paddlefleet/transformer/multi_latent_attention.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/multi_latent_attention.py:2317]: 新提交增加了 `assert get_context_parallel_world_size() == 1`，但 Python 在 `-O` 模式下会移除该检查，CP decode 仍会进入此前指出的 local/global `doc_valid_range` 坐标不一致路径。请改为显式 `ValueError`（或完成 CP 坐标修复）；否则该 P1 仍未解决。

Comment (risemeup1111) [reply src/paddlefleet/transformer/multi_latent_attention.py:2317]: 当前提交已将该 `assert` 改为显式 `ValueError`，并新增 CP>1 decode 拒绝用例；`python -O` 下保护不会再失效。本项已解决。

**Review by risemeup1111 — APPROVED** [review]
结论：APPROVE（当前 head `37977a6f`）。

已复核当前 head 相对 base `e8f705b2` 的 9 个文件及全部历史线程：原 P0 已由 full-layer decode 的 `_hy_sparse_decode_block_indices`、累计 `shared_key` 和 SWA block-sparse 分支恢复；原 P1 已在 `src/paddlefleet/transformer/multi_latent_attention.py:2312` 通过显式 `ValueError` 拒绝 CP>1 decode，并由新增用例覆盖，`python -O` 下保护仍有效。新增 backend-free decode 单测、cache/no-cache parity 和 sink parity 测试已纳入当前 diff。

验证：`compileall` 与 `git diff --check` 通过；本环境未安装 Paddle，未宣称 runtime/GPU 测试通过。标题格式仍是低优先级建议，不构成代码阻塞。

**Review by changeyoung98 — APPROVED** [review]
